### PR TITLE
Merge filter-based testing into webapp

### PR DESCRIPTION
**Pull request by mistzzt**
Filter-testing now works on the web interface; merged #44 into `webapp_fix` branch and fixed some minor issues.

Some description about the fix in case of insterested:
- Previously _webapp_ read from the console output from _hplus_, but with the assumption that each line can only be either solution or stars. Now we request _webapp_ to explicitly find some line in the format `SOLUTION: xxxxx`.

### CODE REVIEW

**Comment by michaelbjames** [comment]
- We definitely still want to have a `--disable-filter` flag. It's useful for when we want to make speed or quality comparisons.
- Did you try this with the `--threaded` flag in the `package.yaml` file to see if we can get multithreading with the webapp?

**Comment by mistzzt** [comment]
> We definitely still want to have a --disable-filter flag. It's useful for when we want to make speed or quality comparisons.

Sorry for the confusion; I didn't mean that the option should be removed.
Instead, previously the webapp run hplus utility with `--disable-filter` flag by default, and I removed it as follows:
```haskell
 proc <- spawnCommand $ printf "stack exec -- hplus \"%s\" --stop-refine --stop-threshold=10 --cnt=10 > %s" query uuid
```

> Did you try this with the --threaded flag in the package.yaml file to see if we can get multithreading with the webapp?

Haven't yet. Will post results here once done.

**Comment by mistzzt** [comment]
I saw that `-threaded` is used with the webapp, and it works fine. Is this what you mean by _multithreading with the webapp_? @michaelbjames

**Comment by michaelbjames** [comment]
What I mean is actual multithreading between the synthesis and the filtering threads within one call to `hplus --<args>`

**Comment by michaelbjames** [comment]
So I tried compiling with `-threaded -rtsopts=-N` and got a segfault on the query `(a->b) -> a -> (a, b)`.
On the query: `(a->b, a) -> b` I got a z3 error.

Suffice to say I don't think we're ready to turn on proper multi-threading support in the hplus application (separate from the webapp, which already supports many simultaneous connections).

**Comment by michaelbjames** [comment]
So, @mistzzt what's been the slowest query you've seen, or where your filtering tends to take the longest? I'm curious to see how much overhead this ends up adding.

**Comment by mistzzt** [comment]
> what's been the slowest query you've seen, or where your filtering tends to take the longest? I'm curious to see how much overhead this ends up adding.

Theoretically, each individual execution of a solution takes `5 * 10^4` microseconds; we typically run 100 tests for each task thus the total will be `5 * 10^6` ms = 5 seconds. However, I've noticed that the performance was sometimes affected and takes about 20 seconds for all tasks to finish. (maybe it's an issue about concurrency?)

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "name": "H+ w/ HIE(G) 0.11",
+    "image": "mistzzt/hoogle_plus_dev:latest",
+
+    "extensions": [
+        "eamodio.gitlens",
+        "alanz.vscode-hie-server",
+        "aaron-bond.better-comments"
+    ]
+}
```
Comment (michaelbjames) [inline .devcontainer/devcontainer.json]: Thanks for including this. Would you mind updating the README's section about how to setup a decent docker development flow?

File: docker-compose.sh [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+sudo docker-compose up --scale hplus=4
```
Comment (michaelbjames) [inline docker-compose.sh]: We don't use this anymore, so would you delete this and docker-compose.yaml files?

**Comment by michaelbjames** [comment]
This looks great! Thanks for making those changes.
Go ahead and Squash+Merge at your discretion.

**Pull request merged**